File: modules/core/src/main/java/com/eventhorizon/pulsar/core/packet/handler/PacketHandler.java
package com.eventhorizon.pulsar.core.packet.handler;

import com.eventhorizon.pulsar.core.packet.Packet;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

@Getter
@RequiredArgsConstructor
public abstract class PacketHandler<T extends Packet> {

    private final PacketHandlerPriority priority;

    public void onSendPacket(T packet) {}

    public void onReceivePacket(T packet) {}
}

File: modules/core/src/main/kotlin/com/eventhorizon/pulsar/core/packet/handler/PacketHandler.kt
package com.eventhorizon.pulsar.core.packet.handler

import com.eventhorizon.pulsar.core.packet.Packet

enum class PacketHandlerPriority {
    LOWEST,
    LOW,
    NORMAL,
    HIGH,
    HIGHEST;
}

abstract class PacketHandler<T : Packet> {
    open val priority: PacketHandlerPriority = PacketHandlerPriority.NORMAL

    open fun onSendPacket(packet: T) {}
    open fun onReceivePacket(packet: T) {}
}
